sklearn/kernel_approximation.py

```diff
@@ -22,6 +24,168 @@
+
+    Implements Tensor Sketch, which approximates the feature map
+    of the polynomial kernel::
+
+        K(X, Y) = (gamma * <X, Y> + coef0)^degree
+
+    by efficiently computing a Count Sketch of the outer product of a
+    vector with itself. Read more in the
+    :ref:`User Guide <polynomial_kernel_approx>`.
+
+    Parameters
+    ----------
+    gamma : float
+        Parameter of the polynomial kernel whose feature map
+        will be approximated.
+
+    degree : int
+        Degree of the polynomial kernel whose feature map
+        will be approximated.
+
+    coef0 : int
+        Constant term of the polynomial kernel whose feature map
+        will be approximated.
+
+    n_components : int
+        Dimensionality of the output feature space.
+
+    random_state : int, RandomState instance, default=None
+        Determines random number generation for indexHash and bitHash
+        initialization. Pass an int for reproducible results across multiple
+        function calls. See :term:`Glossary <random_state>`.
+
+    Attributes
+    ----------
+    indexHash_ : ndarray of shape (degree, n_features), dtype=int64
+        Array of indexes in range [0, n_components) used to represent
+        the 2-wise independent hash functions for Count Sketch computation.
+
+    bitHash_ : ndarray of shape (degree, n_features), dtype=float32
+        Array with random entries in {+1, -1}, used to represent
+        the 2-wise independent hash functions for Count Sketch computation.
+
+    Examples
+    --------
+    >>> from sklearn.kernel_approximation import PolynomialSampler
+    >>> from sklearn.linear_model import SGDClassifier
+    >>> X = [[0, 0], [1, 1], [1, 0], [0, 1]]
+    >>> y = [0, 0, 1, 1]
+    >>> ps = PolynomialSampler(degree=3, random_state=1)
+    >>> X_features = ps.fit_transform(X)
+    >>> clf = SGDClassifier(max_iter=10, tol=1e-3)
+    >>> clf.fit(X_features, y)
+    SGDClassifier(max_iter=10)
+    >>> clf.score(X_features, y)
+    1.0
+    """
+
+    def __init__(self, *, gamma=1., degree=2, coef0=0, n_components=100,
+                 random_state=None):
+
+        self.gamma = gamma
+        self.degree = degree
+        self.coef0 = coef0
+        self.n_components = n_components
+        self.random_state = random_state
+
+    def fit(self, X, y=None):
+        """Fit the model with X.
+
+        Initializes the internal variables. The method needs no information
+        about the distribution of data, so we only care about n_features in X.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            Training data, where n_samples in the number of samples
+            and n_features is the number of features.
+
+        Returns
+        -------
+        self : object
+            Returns the transformer.
+        """
+
+        X = check_array(X, accept_sparse="csc")
+        random_state = check_random_state(self.random_state)
```

Comment:
The issue is that calling fit twice would produce a different seed and therefore a different result, since RandomState instance is mutable. In https://github.com/scikit-learn/scikit-learn/pull/14605 for a similar use case, we added a `seed` variable in transform, but I'm not particularly happy with that outcome either. This is probably fine as is, we would just have to address this globally at some point in https://github.com/scikit-learn/scikit-learn/issues/14042